src/zkevm_specs/evm/execution/mul_div_mod.py

```diff
@@ -0,0 +1,71 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ...util import FQ
+
+
+def mul_div_mod(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    # The opcode value for MUL, DIV and MOD is 2, 4, 6. When the opcode is MUL,
+    # (Opcode.DIV - opcode) * (Opcode.MOD - opcode) is 8. To make `is_mul` be
+    # either 0 or 1, we need to divide the product by 8, which is equivalent to
+    # multiply it by inversion of 8. Similarly, we also need to multiply the
+    # inversion of 4 and 8 for `is_div` and `is_mod` respectively.
+    is_mul = (Opcode.DIV - opcode) * (Opcode.MOD - opcode) * FQ(8).inv()
+    is_div = (opcode - Opcode.MUL) * (Opcode.MOD - opcode) * FQ(4).inv()
+    is_mod = (opcode - Opcode.MUL) * (opcode - Opcode.DIV) * FQ(8).inv()
+
+    pop1 = instruction.stack_pop()
+    pop2 = instruction.stack_pop()
+    push = instruction.stack_push()
+
+    # this part corresponds to witness assignment in the zkevm circuit
+    if is_mul == 1:
+        a = pop1
+        b = pop2
+        c = instruction.rlc_encode(0, 32)
+        d = push
+    elif is_div == 1:
+        d = pop1  # dividend
+        b = pop2  # divisor
+        a = push  # quotient
+        c = instruction.rlc_encode(d.int_value - b.int_value * a.int_value, 32)  # remainder
+    else:  # is_mod == 1
+        d = pop1  # dividend
+        b = pop2  # divisor
+        if b.int_value == 0:
+            c = d
+            a = instruction.rlc_encode(0, 32)
+        else:
+            c = push
+            a = instruction.rlc_encode((d.int_value - c.int_value) // b.int_value, 32)
+
+    divisor_is_zero = instruction.word_is_zero(b)
+    overflow = instruction.mul_add_words(a, b, c, d)
+
+    # constrain the push and pop values
+    instruction.constrain_equal(pop1, instruction.select(is_mul, a, d))
+    instruction.constrain_equal(pop2, b)
+    instruction.constrain_equal(
+        push,
+        is_mul * d.expr()
+        + is_div * a.expr() * (1 - divisor_is_zero)
+        + is_mod * c.expr() * (1 - divisor_is_zero),
+    )
+
+    # constrain c == 0 for MUL
+    instruction.constrain_zero(is_mul * instruction.sum(c.le_bytes))
+
+    # constrain remainder < divisor when divisor != 0 for DIV and MOD
+    lt, _ = instruction.compare_word(c, b)
+    instruction.constrain_zero((1 - is_mul) * (1 - divisor_is_zero) * lt)
```

Comment:
when we have "not mul" or "divisor not zero", the c < b. So the "lt" would be true. 
The expression `(1 - is_mul) * (1 - divisor_is_zero) * lt` evaluates to `lt`, which would be 1 not 0, failing the constrain_zero check.

```suggestion
    instruction.constrain_zero((1 - is_mul) * (1 - divisor_is_zero) * (1 - lt))
```

Reply:
After fixing that, the only test failing is this. I posted the full output since this is a randomized one.
```
============================================================ FAILURES ============================================================
_ test_mul_div_mod[Opcode.MOD-42390196743033146959778914371328500883827744724379697650640924689959794134393-102270312980869441008273231372570788922717343670179665268455894392667877226873] _

opcode = <Opcode.MOD: 6>, a = RLC(42390196743033146959778914371328500883827744724379697650640924689959794134393)
b = RLC(102270312980869441008273231372570788922717343670179665268455894392667877226873)

    @pytest.mark.parametrize("opcode, a, b", TESTING_DATA)
    def test_mul_div_mod(opcode: Opcode, a: int, b: int):
        randomness = rand_fq()
    
        if opcode == Opcode.MUL:
            c = a * b % 2**256
        elif opcode == Opcode.DIV:
            c = 0 if b == 0 else a // b
        else:  # Opcode.MOD
            c = 0 if b == 0 else a % b
    
        a = RLC(a, randomness)
        b = RLC(b, randomness)
        c = RLC(c, randomness)
    
        if opcode == Opcode.MUL:
            bytecode = Bytecode().mul(a, b)
        elif opcode == Opcode.DIV:
            bytecode = Bytecode().div(a, b)
        else:
            bytecode = Bytecode().mod(a, b)
        bytecode_hash = RLC(bytecode.hash(), randomness)
    
        tables = Tables(
            block_table=set(Block().table_assignments(randomness)),
            tx_table=set(),
            bytecode_table=set(bytecode.table_assignments(randomness)),
            rw_table=set(
                RWDictionary(9)
                .stack_read(1, 1022, a)
                .stack_read(1, 1023, b)
                .stack_write(1, 1023, c)
                .rws
            ),
        )
    
>       verify_steps(
            randomness=randomness,
            tables=tables,
            steps=[
                StepState(
                    execution_state=ExecutionState.MUL,
                    rw_counter=9,
                    call_id=1,
                    is_root=True,
                    is_create=False,
                    code_source=bytecode_hash,
                    program_counter=66,
                    stack_pointer=1022,
                    gas_left=5,
                ),
                StepState(
                    execution_state=ExecutionState.STOP,
                    rw_counter=12,
                    call_id=1,
                    is_root=True,
                    is_create=False,
                    code_source=bytecode_hash,
                    program_counter=67,
                    stack_pointer=1023,
                    gas_left=0,
                ),
            ],
        )

tests/evm/test_mul_div_mod.py:82: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
src/zkevm_specs/evm/main.py:21: in verify_steps
    verify_step(
src/zkevm_specs/evm/main.py:38: in verify_step
    EXECUTION_STATE_IMPL[instruction.curr.execution_state](instruction)
src/zkevm_specs/evm/execution/mul_div_mod.py:62: in mul_div_mod
    instruction.constrain_zero((1 - is_mul) * (1 - divisor_is_zero) * (1-lt))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <zkevm_specs.evm.instruction.Instruction object at 0x10c83f6d0>, value = 1

    def constrain_zero(self, value: Expression):
>       assert value.expr() == 0, ConstraintUnsatFailure(f"Expected value to be 0, but got {value}")
E       AssertionError: Expected value to be 0, but got 1

src/zkevm_specs/evm/instruction.py:124: AssertionError
------------------------------------------------------ Captured stdout call ------------------------------------------------------
compare_word 0 RLC(42390196743033146959778914371328500883827744724379697650640924689959794134393) RLC(102270312980869441008273231372570788922717343670179665268455894392667877226873)
==================================================== short test summary info =====================================================
FAILED tests/evm/test_mul_div_mod.py::test_mul_div_mod[Opcode.MOD-42390196743033146959778914371328500883827744724379697650640924689959794134393-102270312980869441008273231372570788922717343670179665268455894392667877226873]
================================================== 1 failed, 76 passed in 1.51s ==================================================

```

Reply:
Yes, fixed it. It was still a bug in the `compare_word`. now all tests passed.